sparkmagic/proposal.md

```diff
@@ -0,0 +1,49 @@
+# Proposal for Incubation
+
+## Subproject name
+
+sparkmagic
+
+## Development team and Advocate
+
+Subproject development team:
+
+* Alejandro Guerrero Gonzalez, Microsoft (`@aggFTW`)
+* Joel Zambrano, Microsoft (`@djoelz`)
+* Auberon Lopez, Cal Poly (`@alope107`)
+
+Steering Council Advocate
+
+* Brian Granger (`@ellisonbg`)
+
+## Subproject goals, scope and functionality
+
+Goals:
+* Make Spark usage from the IPython notebook as frictionless as possible by decoupling Spark code execution from Spark installation.
+* These magics will be packaged so that they can be pip-installable or conda-installable.
+* Allow any Jupyter installation to point to different Spark installations.
+
+Scope:
+* Spark will be the first back end provided by these magics, but it could easily be extended to other big data back ends.
+* Remote Spark execution through a REST endpoint which allows for Scala, Python, and R support as of Sept 2015.
+
+Functionality:
+* Enable users to remotely connect to a REST Spark back end.
+* Return of pandas dataframes from Spark when appropriate. This will allow users to integrate spark with the Python libraries available for pandas dataframes.
+
+## Audience
+
+Spark users who want to have the best experience possible from a Notebook. These users care about the libraries ecosystem the Jupyter community has created.
+
+## Other options
+
+Alternatives we know of:
+
+* Combination of IPython, R kernel (or rpy2), and Scala kernel for an in-Spark-cluster Jupyter installation. This does not allow the user to point to different Spark clusters. It might also result in resource contention (CPU or memory) between the Jupyter installation and Spark.
+* IBM's Spark kernel does not provide a REST endpoint, requires the installation of Jupyter in the cluster, and does not create pandas dataframes.
```

Comment:
@freeman-lab No worries. I was adding the clarification for others to make sure we're all on the same page. I can tell you're "Spark initiated". ;)

> One thing I'm curious about for the RESTful model would be how large aggregations behave, i.e. when a large result needs to be returned to the driver. Do you have experience with this? Maybe that's what you already meant by

Sending a large (or even medium-sized) result to a driver is definitely one of the cases I was talking about. @lbustelo is stating another down in the comments below, namely the number of protocols / serializations involved in:

```
notebook frontend --Jupyter protocol over websocket--> 
notebook backend --Jupyter protocol over 0mq--> 
Python kernel --Proposed protocol over HTTP--> 
proposed REST API --??? protocol (Livy?) over ???--> 
some local process in language of choice (e.g., provided by livy?) --Spark protocols (Akka, etc.)--> 
Spark executor```

and back. Versus

```
notebook frontend --Jupyter protocol over Websocket-->  
notebook backend --Jupyter protocol over websocket--> 
kernel gateway --Jupyter protocol over 0mq--> 
Scala, R, or Python kernel --> --Spark protocols (Akka, etc.)--> 
Spark executor
```

in the kernel provisioner / gateway proposal that we clearly need to submit as an incubator project real-soon-now so everyone can start to understand the pros/cons of all these approaches. :)